### PR TITLE
feat: support must change password option for create user

### LINKED ISSUE #15956: Feature:  support create user and alter user  option   MUST_CHANGE_PASSWORD = TRUE
**Summary**

create user |alter user need user login first change password.

The user password was expired login, first must change the password.

> https://docs.snowflake.com/en/sql-reference/sql/create-user

### PR DESCRIPTION
**Pull request by b41sh**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

When create or alter a user, we can set an option `must_change_password` as true. The new user cannot execute any operations after login for the first time and must change the password first.

In addition, if the password policy is set, After the password is changed within the maximum time limit, the user can still login. However, the user must change the password before execute other operations.

for example

```sql
➜  ~ mysql -uroot -P3307 -h0.0.0.0 -A
Welcome to the MySQL monitor.  Commands end with ; or \g.

# admin create a new user and set option `must_change_password` as true
mysql> create user user1 identified by 'abc123' with must_change_password = true;
Query OK, 0 rows affected (0.05 sec)

mysql> exit

# new user login with the password set by admin
➜  ~ mysql -uuser1 -pabc123 -P3307 -h0.0.0.0 -A
Welcome to the MySQL monitor.  Commands end with ; or \g.

# new user can't do any operations except change password
mysql> select 123;
ERROR 1105 (HY000): PermissionDenied. Code: 1063, Text = Must change password before execute other operations.

mysql> alter user user() identified by 'abc456';
Query OK, 0 rows affected (0.03 sec)

mysql> select 123;
ERROR 1105 (HY000): PermissionDenied. Code: 1063, Text = Must change password before execute other operations.
mysql> exit

# new user login with the changed password
➜  ~ mysql -uuser1 -pabc456 -P3307 -h0.0.0.0 -A
Welcome to the MySQL monitor.  Commands end with ; or \g.

# new user can do any operations
mysql> select 123;
+------+
| 123  |
+------+
|  123 |
+------+
1 row in set (0.40 sec)
Read 1 rows, 1.00 B in 0.376 sec., 2.66 rows/sec., 2.66 B/sec.
```

- Fixes #15956

## Tests

- [x] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16031)

### CODE REVIEW

File: src/query/users/src/password_policy.rs [hunk 1]
```diff
@@ -431,22 +431,21 @@ impl UserApiProvider {
...
-                            max_change_time
-                        );
-                        // Password has not been changed for more than max age days, cannot login
+                        // Password has not been changed for more than max age days,
+                        // allow login, but must change password first.
                         if let Ordering::Less = max_change_time.cmp(&now) {
-                            return Err(ErrorCode::InvalidPassword(
-                                "Password has not been changed more than max age days".to_string(),
-                            ));
+                            info!(
```
Comment (TCeason) [inline src/query/users/src/password_policy.rs:437]: warn level?

File: src/query/service/src/interpreters/access/privilege_access.rs [hunk 2]
```diff
@@ -566,6 +566,18 @@ impl AccessChecker for PrivilegeAccess {
...
+        // if user need change password, only change self password is allowed.
+        let need_change = user.auth_info.get_need_change();
+        if need_change {
+            if let Plan::AlterUser(plan) = plan {
+                if plan.user.username == user.name && plan.user_option.is_none() {
+                    return Ok(());
+                }
+            }
+            return Err(ErrorCode::PermissionDenied(
+                "Must change password before execute other operations".to_string(),
```
Comment (TCeason) [inline src/query/service/src/interpreters/access/privilege_access.rs:578]: If MUST_CHANGE_PASSWORD=true is set, also support login but need to modify password first?

File: tests/suites/0_stateless/18_rbac/18_0006_mysql_auth.py [hunk 3]
```diff
@@ -35,6 +36,9 @@
     cursor.execute(
         "create user u3 identified by 'abc123' with set network policy='p2';"
     )
+    cursor.execute(
+        "create user u4 identified by 'abc123' with must_change_password = true;"
```
Comment (TCeason) [inline tests/suites/0_stateless/18_rbac/18_0006_mysql_auth.py:40]: Better to mock a password failed test? e.g.

create user u4 identified by 'abc123' with must_change_password = false;

Bound a password policy that max_failed_connect 3 times to the user u5, after failed 3 time, the u5 can be login with correct password. But only support modify password?

**Review by drmingdrmer — APPROVED** [review]
> Reviewed 20 of 33 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16031)* status: 20 of 33 files reviewed, 3 unresolved discussions (waiting on @b41sh and @sundy-li)

File: src/query/service/src/interpreters/access/privilege_access.rs (hunk 2 above)
Comment (TCeason) [reply src/query/service/src/interpreters/access/privilege_access.rs:578]: Maybe use other error code?

Comment (b41sh) [reply src/query/service/src/interpreters/access/privilege_access.rs:609]: add a new error code `NeedChangePasswordDenied`

File: tests/suites/0_stateless/18_rbac/18_0006_mysql_auth.py (hunk 3 above)
Comment (b41sh) [reply tests/suites/0_stateless/18_rbac/18_0006_mysql_auth.py:40]: the default value of `must_change_password` is false, `u1`, `u2` and `u3` is this case, we don't need add a new user.
the min value of `max_age_days` in password policy is 1 day, it means we must wait 1 day to test this case, it not easy to test here.

File: src/query/service/src/interpreters/access/privilege_access.rs [hunk 4]
```diff
@@ -574,6 +574,31 @@ impl AccessChecker for PrivilegeAccess {
...
+                let mut allowed = true;
+                let metadata = metadata.read().clone();
+                for table in metadata.tables() {
+                    let db_name = table.database();
+                    let table_name = table.name();
+                    if !((db_name == "system"
+                        && SYSTEM_TABLES_ALLOW_LIST.iter().any(|x| x == &table_name))
+                        || db_name == "information_schema")
+                    {
+                        allowed = false;
```
Comment (TCeason) [inline src/query/service/src/interpreters/access/privilege_access.rs:599]: cc @wubx Please review this. MySQL user can do simple select?

Comment (b41sh) [reply src/query/service/src/interpreters/access/privilege_access.rs:599]:
```sql
SELECT version()

select name from system.tables union all select name from system.columns union all select name from system.databases union all select name from system.functions
```

When using `BendSQL` client to login, those two SQL will be automatically executed. If the privilege check fails here, `BendSQL` will be panic, and user can't change his password.

Comment (b41sh) [reply src/query/service/src/interpreters/access/privilege_access.rs]: have delete this logic, we modify `BendSQL` to fix the problem

**Review by TCeason — APPROVED** [review]

**Comment by wubx** [comment]
@soyeric128  This feature need doc. 
Put it at create user  and alter user

**Pull request merged**